src/kbmod/work_unit.py

```diff
@@ -0,0 +1,170 @@
+from astropy.io import fits
+from astropy.table import Table
+import numpy as np
+from pathlib import Path
+
+from kbmod.configuration import SearchConfiguration
+from kbmod.search import ImageStack, LayeredImage, PSF, RawImage
+
+
+class WorkUnit:
+    """The work unit is a storage and I/O class for all of the data
+    needed for a full run of KBMOD, including the: the search parameters,
+    data files, and the data provenance metadata.
+
+    A WorkUnit file is a FITS file with the following extensions:
+        0 - Primary header with overall metadata
+        1 - The data provenance metadata
+        2 - The search parameters
+        3 through X - Image layers with either alternating layers of science, variance,
+            mask, and PSF. Layers may be empty if no data is provided.
+    """
+
+    def __init__(self, im_stack=None, config=None):
+        self.im_stack = im_stack
+        self.config = config
+
+    @classmethod
+    def from_file(cls, filename):
+        """Create a WorkUnit from a single file.
+
+        Parameters
+        ----------
+        filename : `str`
+            The file to load.
+
+        Returns
+        -------
+        result : `WorkUnit`
+            The loaded WorkUnit.
+        """
+        if not Path(filename).is_file():
+            raise ValueError(f"WorkUnit file {filename} not found.")
+
+        result = None
+        with fits.open(filename) as hdul:
+            num_layers = len(hdul)
+            if num_layers < 5:
+                raise ValueError(f"WorkUnit file has too few extensions {len(hdul)}.")
+
+            # TODO - Read in provenance metadata from extension #1.
+
+            # Read in the search parameters from extension #2.
+            config = SearchConfiguration()
+            config_table = Table(hdul[2].data)
+            config.set_from_table(config_table)
+
+            # Read the size and order information from the primary header.
+            num_images = hdul[0].header["NUMIMG"]
+            if len(hdul) != 4 * num_images + 3:
+                raise ValueError(
+                    f"WorkUnit wrong number of extensions. Expected "
+                    f"{4 * num_images + 3}. Found {len(hdul)}."
+                )
+
+            # Read in all the image files.
+            imgs = []
+            for i in range(num_images):
+                ext_num = 3 + 4 * i
+
+                # Read in science and variance layers.
+                sci = hdu_to_raw_image(hdul[ext_num])
+                var = hdu_to_raw_image(hdul[ext_num + 1])
+
+                # Read the mask layer if it exists.
+                msk = hdu_to_raw_image(hdul[ext_num + 2])
+                if msk is None:
+                    msk = RawImage(np.zeros((sci.get_height(), sci.get_width())))
+
+                # Check if the PSF layer exists.
+                if hdul[ext_num + 3].header["NAXIS"] == 2:
+                    p = PSF(hdul[ext_num + 3].data)
+                else:
+                    p = PSF(1e-8)
```

Comment:
We are the sole creator of these files. I think we should just throw an error if we don't have everything we need before writing it to begin with. Or make these defaults the matter of RawImage/PSF constructors or something and write whatever they make to disk.